wasm/src/lsp/types.rs

```diff
@@ -352,3 +355,136 @@ fn test_doc_range_overlap_at_zero() {
         end: DocPosition { line: 0, character: 3 }
     }));
 }
+
+// An object that contains the literal text of a document we're working with in
+// the LSP.
+#[derive(Debug, Clone)]
+pub struct DocData {
+    pub fullname: String,
+    pub text: Vec<Rc<Vec<u8>>>,
+    pub version: i32,
+    // Zero based.
+    pub comments: HashMap<usize, usize>,
+}
+
+impl DocData {
+    // Return a reference to the nth line's data.
+    pub fn nth_line_ref(&self, line: usize) -> Option<&Vec<u8>> {
+        if line < self.text.len() {
+            let borrowed: &Vec<u8> = self.text[line].borrow();
+            Some(borrowed)
+        } else {
+            None
+        }
+    }
+
+    // Given a position go back one character, returning the character
+    // and the new position if they exist.
+    pub fn get_prev_position(&self, position: &Position) -> Option<(u8, Position)> {
+        if position.character == 0
+            && position.line > 0
+            && ((position.line - 1) as usize) < self.text.len()
+        {
+            let nextline = position.line - 1;
+            self.get_prev_position(&Position {
+                line: nextline,
+                character: self.text[nextline as usize].len() as u32,
+            })
```

Comment:
When
```rust
let dd = DocData {
    ...,
    text: vec![
        vr("(mod (X)"),
        vr(""),
        vr(""),
        vr(""),
        vr(" (defun F (A) (+ A 1)) ;; A function"),
        vr(" (F X) ;; Call"),
        vr(" )")
    ]
}
```
and
```rust
let result = dd.get_prev_position(&Position { line: 4, character: 0 });
```
the returned `result.unwrap().1` would be `Position {line: 0, character: 7}`.
Is this expected behavior?
Meaning, empty lines should be skipped?

Reply:
yes, in the future this past is part of, it will have been used here:

    // A position points to "first in list" if reversing past all alphanumerics
   // then all spaces yields a character other than '('.
   pub fn is_first_in_list(lines: &DocData, position: &Position) -> bool {

to find an opening paren that encloses the chosen position, so it's ok to skip blank lines.

I'll give it a comment to that effect.